src/polydatum/dal.py

```diff
@@ -1,20 +1,55 @@
 from __future__ import absolute_import
+
+import inspect
 from contextlib import contextmanager
-from .context import DataAccessContext
-from polydatum.errors import AlreadyExistsException
+
+from polydatum.middleware import PathSegment, DalMethodRequest, DeferredDalAttrRequest, dal_resolver, \
+    dal_method_resolver_middleware, handle_dal_method
+from polydatum.errors import AlreadyExistsException, InvalidMiddleware
 from polydatum.util import is_generator
 from .resources import ResourceManager
 from .context import _ctx_stack
+from typing import Callable, Tuple
+from functools import partial, update_wrapper
+from polydatum.context import DataAccessContext
 
 
 class DataAccessLayer(object):
     """
     Gives you access to a DataManager's services.
     """
 
-    def __init__(self, data_manager):
+    # default middleware classes need to be instantiated before being
+    # passed to the init function because the resulting object is called
+    # directly (__call__).
+    def __init__(
+        self,
+        data_manager,
+        middleware=None,
+        default_middleware=(dal_method_resolver_middleware,),
+        handler=handle_dal_method
+    ):
         self._services = {}
         self._data_manager = data_manager
+        self._handler = handler
+        self._reversed_middleware = []
```

Comment:
We still don't need `self._reversed_middleware` or the test that tests it directly. There's a good test you provided that tests order much better by checking only input/output and doesn't worry about actual middleware internals.

Reply:
The only retort I can come up with here is to not think of the test as verifying package functionality but more the test is verifying the author's understanding of what is happening. If you want it removed we can remove the test and instance variable (just use a local var).

Reply:
Ideally, the author should only test input/output. If it's not clear how to do that, then a deeper understanding of how it is actually working is needed. If you truly understand how to test it via input/output only, then you have a deep understanding of the internals.